src/state/selectors.ts

```diff
@@ -26,29 +31,60 @@ function getSliced<T>(arr: ReadonlyArray<T>, startIndex: number, endIndex: numbe
-  const { activeIndex, tooltipItemPayloads } = state.tooltip;
-  if (activeIndex === -1) {
-    return undefined;
-  }
-  const { chartData, dataStartIndex, dataEndIndex } = state.chartData;
-  /*
-   * If a payload has data specified directly from the graphical item, prefer that.
-   * Otherwise, fill in data from the chart level, using the same index.
-   */
-  return tooltipItemPayloads.map(({ dataDefinedOnItem, settings }) => {
-    const finalData = dataDefinedOnItem ?? chartData;
-
-    const sliced = getSliced(finalData, dataStartIndex, dataEndIndex);
-
-    // TODO settings coming from Tooltip props, and tooltipAxis
-    const tooltipPayload = sliced?.[activeIndex];
-    const tooltipEntry = getTooltipEntry({
-      tooltipEntrySettings: settings,
-      dataKey: settings?.dataKey,
-      payload: tooltipPayload,
-      value: getValueByDataKey(tooltipPayload, settings?.dataKey),
+const selectTooltipState = (state: RechartsRootState) => state.tooltip;
+const selectChartData = (state: RechartsRootState) => state.chartData;
+
+const selectTooltipTicks = createSelector(selectTooltipAxis, (tooltipAxis: BaseAxisProps) =>
+  getTicksOfAxis(tooltipAxis, false, true),
+);
+
+const selectActiveIndex = createSelector(selectTooltipState, (tooltipState: TooltipState) => tooltipState.activeIndex);
+
+const selectActiveLabel = createSelector(
+  selectTooltipTicks,
+  selectActiveIndex,
+  (tooltipTicks: ReadonlyArray<TickItem>, activeIndex: number): string | undefined =>
+    tooltipTicks?.[activeIndex]?.value,
+);
+
+export const selectTooltipPayload: (state: RechartsRootState) => TooltipPayload | undefined = createSelector(
+  selectTooltipState,
+  selectChartData,
+  selectTooltipAxis,
+  selectActiveLabel,
+  (
+    tooltipState: TooltipState,
+    chartDataState: ChartDataState,
+    tooltipAxis: BaseAxisProps | undefined,
+    activeLabel: string | undefined,
+  ) => {
+    const { activeIndex, tooltipItemPayloads } = tooltipState;
+    if (activeIndex === -1) {
+      return undefined;
+    }
+    const { chartData, dataStartIndex, dataEndIndex } = chartDataState;
+
+    /*
+     * If a payload has data specified directly from the graphical item, prefer that.
+     * Otherwise, fill in data from the chart level, using the same index.
+     */
+    return tooltipItemPayloads.map(({ dataDefinedOnItem, settings }) => {
+      const finalData = dataDefinedOnItem ?? chartData;
+
+      const sliced = getSliced(finalData, dataStartIndex, dataEndIndex);
+
+      let tooltipPayload: unknown;
+      if (tooltipAxis?.dataKey && !tooltipAxis?.allowDuplicatedCategory) {
+        tooltipPayload = findEntryInArray(sliced, tooltipAxis.dataKey, activeLabel);
+      } else {
+        tooltipPayload = sliced?.[activeIndex];
+      }
+      return getTooltipEntry({
+        tooltipEntrySettings: settings,
+        dataKey: settings?.dataKey,
+        payload: tooltipPayload,
+        value: getValueByDataKey(tooltipPayload, settings?.dataKey),
```

Comment:
I think this also needs to be `settings?.dataKey ?? tooltipAxis?.dataKey`

Else no value will be rendered in tooltip when the axis has the dataKey (at least for Scatter)

Reply:
Also really wonder how Cell is going to work :/

Reply:
Also this is, in the tooltipAxis case, applying the dataKey twice? This looks like it shouldn't work at all